### ISSUE TITLE
✨Universal Syntax Editor

### ISSUE THREAD

**Issue by AndrewRedican** [AndrewRedican/react-json-editor-ajrm#28]
Hey @PatrickSachs,

About the contributions so far, thank you. I saw your online resume, I must say kudos to you, creating ☀️ SolScript. I think it is very impressive. I've shared your project with some work colleges. I would like to extend access to the github repo to you as collaborator, not just as contributor - if you would like.

On a different topic, today I had a epiphany while traveling on the bus.

What if we could make this project's logic extensible? -So that it no longer only checks json or javascript syntax, but instead cover other programing languages, aiming first for the most used ones and perhaps others like Solscript, Lua, etc.

A couple of things have to happen first, like improve stability (add QA tests, fix other bugs), improve workflow (automate some publishing tasks), and other revelant features like collapsible nodes, add some additional webhooks or other applications to facilitate collaboration.

Then I would like to create an article explaining my take on developing syntax logic (granted, it can be improved), and then proposing to other developers from the community to create their own.

We could then specify logic through the component's props, similar to what we did with locale. Granted this is an oversimplification of this mamoth task. It's fun and useful, isn't it?

Let me know your thoughts on both subjects.

Cheers.

**Comment by PatrickSachs** [comment]
Hey @AndrewRedican!

First of all, great props to you for reading through my german website! 👍 Thank you for taking the time.

I'd gladly accept collaborator access to this repository, since I believe that this project is the React JS-object editor(or maybe future syntax editor?) with the most future. Others either have a monolithic bundle size, are only React wrappers(Use something like jQuery with direct DOM manipulation under the hood/are just based on ACE/etc...) or display an overly fancy GUI for editing the data. This Editor is simple in a good way while not loosing any of its power. There may be other good ones, but I just haven't found any.

About your suggestion for supporting other syntax formats: It does indeed sound like a **really fun project**, however it would really depend on the desired scope. 

==============

If you are talking about a syntax editor for actual programming languages: While I love the idea of it, manually writing a parser for every supported language would eventually stop being feasible and maintainable(some languages like JavaScript get regular syntax new additions, other languages like Elixir actually remove syntax, etc...). We'd need some rather powerful and complex backend to handle vastly different programming languages, provide auto completion, etc. I honestly think this is out of scope, but I'm always open for debate.  

While not having explicit support for React monaco(the text editor in VS Code) by Microsoft is a rather well established library in this area: [https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-custom-languages](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-custom-languages)

==============

Another thing would be an editor purely for data structures(JSON, XML, Lua Tables, INI, CSV, ...). While these things could potentially be implemented in a library like monaco or using a full fledged editor for programming languages, they are typically overkill for the job.

We could provide a lightweight, powerful and extensible editor for these data languages. Granted, we would probably have to switch using to a library like [nearly.js](https://nearley.js.org/) for parsing, considering how vastly different some languages are(Up for debate though, parsing data languages "by hand" is my experience typically still a realistic thing to do).

==============

Writing articles and improving workflow steps is always a good thing for various reasons, especially in/for open source projects.

So I'm definitely up for collaboration on this, we'd just have to define the scope/architecture of this beforehand. On that note, I really do like your suggestion for plug-able logic through the props - A modular architecture is always great.

Cheers right back! (And sorry for the wall of text!)

**Comment by AndrewRedican** [comment]
Hey @PatrickSachs,

I am glad to have you as the first collaborator.

I have no intention of making this project a direct competitor of React monaco 😅. The scope of this project I would want it to remain the same: a syntax editor for `data structures` only, that it can be extended to different programming languages and/or data formats in general.

Preferably, I'd like to keep the project self-contained with only using dev dependencies to streamline QA test, distribution builds, and project management in general. However, after I skimmed through [nearly.js](https://nearley.js.org/docs/tokenizers) documentation, I realized that my version "2.0" released several months ago, implemented a `tokenizer` solution that is similar to that of [nearly.js](https://nearley.js.org/docs/tokenizers). I would like to add [nearly.js](https://nearley.js.org/docs/tokenizers) as a dependency if it will make it easier for us to extend this project to support other data structure syntaxes. But first I would want to make myself really familiar with it to gauge the benefits of using it rather than coming up with out own solutions.

That being said, welcome aboard.

I am about to add some milestones to the project. Let me know if you'd like to suggest additional tools/apps for collbaoration communication or just keep using Github for everything.

**Comment by PatrickSachs** [comment]
Thank you!

> Preferably, I'd like to keep the project self-contained with only using dev dependencies to streamline QA test, distribution builds, and project management in general.

The more self contained, the better, I totally agree. It it is also better for bundle size. I've seen enough some helper libraries include something like lodash/underscore which then as a result dramatically increase bundle size.

> However, after I skimmed through nearly.js documentation, I realized that my version "2.0" released several months ago, implemented a tokenizer solution that is similar to that of nearly.js. [...]

There also probably many other tokenizer/parser libraries out there, nearly is pretty much the one I've seen recommended the most. And as you already stated, manually writing the parsers is also an option.

> I am about to add some milestones to the project. Let me know if you'd like to suggest additional tools/apps for collbaoration communication or just keep using Github for everything.

About additional tools/apps: Although I've never used it myself, Slack seem to be very popular among developers. If it's overkill for just two people at the moment is something I can't judge. I personally use Discord but I haven't seen it being used for software projects too much.

Do you have any recommendations/preferences about tools/apps? I don't mind using Github/Discord/trying Slack for communication, but if you already know better tools for the job I'm happy to hear about it.

**Comment by AndrewRedican** [comment]
I am happy to use Slack + Github.
> I've got it on my mobile and use it at work too. It's got some great integrations too.

> Awesome! I think our primary focus should be to start with the modularization by separating the JSON part from the actual editor component. [...]

I noticed you've started adding to milestones of your own. Feel free to add other milestones. We can make one for "universal syntax editor" if you'd like. Another for collapsible nodes.

**Comment by PatrickSachs** [comment]
> I've got Slack on my mobile and use it at work too. It's got some great integrations too. I am happy to use Slack + Github.

Then let's settle on Slack. Feel free to shoot me an invitation once a workspace has been set up.

> I noticed you've started adding tasks to milestones. Feel free to add other milestones too. We can make one for "universal syntax editor" if you'd like, and another for collapsible nodes.

Sounds like a good idea!

**Comment by AndrewRedican** [comment]
I've sent an intive for a group named "**syntaxture**".

**Comment by AndrewRedican** [comment]
Closing this in favor of https://github.com/AndrewRedican/react-json-editor-ajrm/issues/50
